### PR TITLE
Rewrite nes.cpp and restructure things for extensibility

### PR DESCRIPTION
**Pull request by chuckries**
This commit restructures the overall nes object and methods in a way that
should make it easier to write an external game loop around the nes.

1. Add Nes::DoFrame and rewrite the Nes::Run method

I was able to isolate the external calls to check for input and blit the
screen outside of the actual nes frame processing. DoFrame will process
one screens worth of pixel data, so it should be used if writing an
external game loop. Run should be used (with an implementation of IGfx and
IInput) to run the nes by itself, without an external game loop.

2. Continue to isolate all SDL references

SDL will continue to be our defualt implementation for rendering, audio,
and input. If nes is launched from command line on any platform, it should
be using SDL.

The goal is to make the core nes code the most usable by an application
wishing to host the emulator.

### CODE REVIEW

File: nes/input.h [hunk 1]
```diff
@@ -60,11 +19,9 @@ class Input : public IMem
     void Save() {}
     void Load() {}
 
-    InputResult CheckInput();
-
-private:
-    void HandleKeyPress(SDL_Keycode code, bool isDown);
+public:
+    JoypadState State;
```
Comment (plnelson) [inline nes/input.h]: May want to rename from "State" to "Joypad".  We may eventually want to support more input device types.

File: nes/nes.cpp [hunk 2]
```diff
@@ -29,85 +40,77 @@ std::unique_ptr<Nes> Nes::Create(const char* romPath)
...
-void Nes::Run()
-{
-    std::shared_ptr<SdlGfx> gfx = std::make_shared<SdlGfx>(3);
-
-    std::shared_ptr<IMapper> mapper = IMapper::CreateMapper(_rom);
+    auto mapper = IMapper::CreateMapper(rom);
     if (mapper == nullptr)
     {
-        return;
+        printf("Mapper not supported\n");
```
Comment (plnelson) [inline nes/nes.cpp]: Doesn't CreateMapper already show a better error message?

File: nes/new_ppu.cpp [hunk 3]
```diff
@@ -500,10 +498,10 @@ void Ppu::DrawFrame()
 
 #endif
 
-    _gfx->Blit(_screen);
+    //_gfx->Blit(_screen);
```
Comment (plnelson) [inline nes/new_ppu.cpp]: Delete?

File: nes/new_ppu.h [hunk 4]
```diff
@@ -218,19 +220,17 @@ class Ppu : public IMem
...
-    void DrawScanline(u8 x);
+    void DrawScanline(u8 x, u8 screen[]);
     bool GetBackgroundColor(u8& paletteIndex);
     bool GetSpriteColor(u8 x, u8 y, bool backgroundOpaque, u8& paletteIndex, SpritePriority& priority);
     void ProcessSprites();
     void PutPixel(u16 x, u16 y, rgb& pixel);
 
 private:
-    std::shared_ptr<IGfx> _gfx;
+    //std::shared_ptr<IGfx> _gfx;
```
Comment (plnelson) [inline nes/new_ppu.h]: Delete

File: nes/input.h (hunk 1 above)
Comment (chuckries) [inline nes/input.h]: I'm going to leave this as is for now just because I'm pretty sure the design of input is still pretty in flux.

**Pull request merged**